Assemble Cypress test suite according to apps installed in current region

s-pic opened the pull request:
# Context

closes https://github.com/FixMyBerlin/fixmy.platform/issues/263

# Description

* Augments the cypress plugin file to ingore tests in certain pages (`Map`, `KatasterKI`) depending on what `REGION` is defined. We can easily extend this logic later on, e.g. by using a Whitelist instead of the current exclusion.
![image](https://user-images.githubusercontent.com/12786207/76795330-3431fc80-67c9-11ea-92ff-8f4c6154ea0e.png)
* adds a central `env.js` file that invokes dotenv and exports the parsed values or any assigned default values.  I encountered this aproach in https://github.com/w3tecch/express-typescript-boilerplate/blob/develop/src/env.ts and I would propose to further implement this out.
It dedupes code, facilitates handeling env var usage by providing variables an IDE can pick up for intellisense and keeps all the logic to parse the `.env` file in one place
* adds a new script `test:e2e-interactive` that runs `cypress open`. This really helped me to inspect the cypress config used
![image](https://user-images.githubusercontent.com/12786207/76795268-12d11080-67c9-11ea-87ac-06be70f95d32.png)

# How can this be tested?

* create a `.env` file in the project root, add `REGION=bonn`
* run `npm run test:e2e`
* inspect that only relevant tests (those in the `Reports` page) are being executed

# Next Steps 

CI Integration

ciex commented on env.js:
```diff
@@ -0,0 +1,15 @@
+// central place to read and expose env vars and to assign default values
+// TODO: consider move other references of process.env to here
+
+// read env vars
+const result = require('dotenv').config();
+
+if (result.error) {
+  throw result.error
+}
+const env = result.parsed;
+
+module.exports = {
+  region: env.REGION || 'bonn',
+  cypressBrowserWindow: env.CYPRESS_BROWSER_WINDOW
+}
```
Is it possible to invoke dotenv directly in the Cypress code so that this module can be removed and all env vars are accessible from Cypress code directly?

ciex replied:
I am not a fan of this pattern I guess.. It moves env vars from `process.env`, where they are expected, to a module. Then, this module a) doesn't include all env vars available in the main application code and b) changes the names of env vars. Both will be sources of errors in the future.

s-pic replied:
Just a proposal, it can easily be removed.
Just let me underline the PROs of this aproach once more:

1. You have one central file which looks up keys in `process.env`, the knowledge is not cluttered. 
1. We gain Intellisense by explicitly stating those properties in a module instead of assuming global properties.
1. We can assign default values and add documentation to those properties in a central place.

Regarding your points
a) Remaining env vars can easily be added.
b) Names can easily be changed using IDE refactoring tools, easier than when we only rely on an opaque `process.env`.

Please consider this pattern once more, but in the end I am also fine to remove the `env.js`.

ciex replied:
For the reasons that you mention I think we should try and use env vars as little as possible.  Env vars should serve three purposes:

a) select the actual application configuration to be loaded (which is what the `REGION` and `BACKEND` env vars do)
b) allow each developer to adapt the application to their development environment (which is what e.g. `CYPRESS_BROWSER_WINDOW` does)
c) hold application secrets that should not be checked into code

Everything else should go into a config file anyway, so we won't ever have lots and lots of env vars. I can understand the advantages that you listed but I think the downsides weigh heavier in this case. 

By the way, you can probably also configure environment variables in your IDE. In my IDE for example, I can create any number of launch scripts that run the same package.json script but with different env vars. Maybe I don't fully understand the advantage that you get in IntelliSense from defining these variables in a module. Is it possible to gain your pro-point 2) without this module?

s-pic replied:
Hey @ciex , I might be on thin ice here but I'll give it another shot.

I propose to use a little JavaScript that _wraps_ (read + assignment of default values) defined environment variables (no matter how, wether this happens in the IDE, a local console or by a CI script). It is not ment to _replace_ the usage of environment vars. An `env.js` is only a file to
* explicitly declare what env vars we expect
* declare fallback values
* facilitate consumption in our code base (intellisense)

Anyways, I like this pattern because I got to know its advantages while working with https://github.com/w3tecch/express-typescript-boilerplate/blob/develop/src/env.ts,
but I don't want to enforce my taste here. 
I am fine to remove the wrapper.

ciex commented on package.json:
```diff
@@ -14,6 +14,7 @@
     "test:e2e": "cypress run",
     "test:e2e-chrome": "cypress run --browser chrome",
     "test:e2e-dev-server": "start-server-and-test start http://localhost:8080 test:e2e",
+    "test:e2e-interactive": "cypress open",
```
This is why the `cypress` script exists in package.json. You can run 

    $ npm run cypress open

ciex replied:
Did you add this script because your IDE doesn't allow adding parameters to commands easily so it wouldn't be convenient for you to run the above command?

s-pic replied:
I added this to be consistent with what we we are doing in the other lines above, e.g. `"test:e2e": "cypress run"`:
We are wrapping a cypress sub-command to get a couple of benefits:
* We describe what a script does using its name: IMHO `"test:e2e-interactive` speaks more to a developer than `cypress open`, I don't have to look up the docs what the specific subcommand dpes
* We can change the implementation of a specific script later on and keep its interface (the script name) stable
* (last and least): Yep, in my IDE I can open a little window showing all available npm scripts and pick a script only using keyboard strokes, which always saves a little time.

ciex replied:
Then let's keep both entries, this one and the `cypress` script. That let's us support both cli and ide workflows.

s-pic commented:
Sorry @ciex for getting back to you that late, I will try to update my github notifications to not miss PR discussions anymore.

Please have a look at the conversation above.

ciex commented:
@s-pic we can go with the whitelisting approach here, which fits our workflow more closely

s-pic commented:
@ciex I 
* removed the `env.js`
* added a whitelist config file `cypressWhiteList.js` (namings and location is only a proposal)
* adapted `cypress/plugins/index` to use the whitelist to determine the pool of test files when cypress boots
* cleaned up the plugin file a bit by factoring out the new logic
* added docs

https://github.com/FixMyBerlin/fixmy.frontend/pull/343/commits/5163f9341fe5def334af03ae249531511ff0d6a4

The functionality can easily be tested by setting a city config and look for the output of added log statements.

![image](https://user-images.githubusercontent.com/12786207/83952217-908c4180-a837-11ea-9c33-eb44be18300f.png)

https://github.com/FixMyBerlin/fixmy.frontend/pull/343/commits/5163f9341fe5def334af03ae249531511ff0d6a4

![image](https://user-images.githubusercontent.com/12786207/83952219-94b85f00-a837-11ea-8457-d693ba2c91a0.png)
